### PR TITLE
execute_with_retry() error handling improvements [VS-159]

### PR DESCRIPTION
**Pull request by rsasch**
Closes https://broadworkbench.atlassian.net/browse/VS-159

- move execute_with_retry to utils so that only defined in one place
- add more specific error handling

**NOTE for reviewers**: you probably want to check "Hide whitespace changes" under settings because IntelliJ did some "helpful" cleanup when I made some code changes to files.

### CODE REVIEW

File: scripts/variantstore/wdl/extract/utils.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            print(f"STARTING - {label} (jobid: {query.job_id})")
+            results = query.result()
+            job = client.get_job(query.job_id)
+            mb_billed = int(0 if job.total_bytes_billed is None else job.total_bytes_billed) / (
+                        1024 * 1024)
+            print(
+                f"COMPLETED ({time.time() - start} seconds, {3 - len(retry_delay)} retries, {mb_billed} MBs) - {label}")
+
+            return results
+        except google.api_core.exceptions.ServiceUnavailable as err:
```
Comment (kcibul) [inline scripts/variantstore/wdl/extract/utils.py:39]: I think you can combine these into a single except block.. see https://stackoverflow.com/questions/6470428/catch-multiple-exceptions-in-one-line-except-block

File: scripts/variantstore/wdl/extract/utils.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import time
+import google.api_core.exceptions
+from google.cloud import bigquery
+
+
+def execute_with_retry(client, label, sql):
+    """Run a BigQuery SQL string with a label.
+
+    Three retries with incremental backoff if BiqQuery returns a 503, any other error is re-raised.
```
Comment (kcibul) [inline scripts/variantstore/wdl/extract/utils.py:9]: Also for BadRequest right (looking at the code below?)

File: scripts/variantstore/wdl/extract/raw_array_cohort_extract.py [hunk 3]
```diff
@@ -1,13 +1,10 @@
 # -*- coding: utf-8 -*-
-import sys
-import uuid
-import time
+import argparse
```
Comment (kcibul) [inline scripts/variantstore/wdl/extract/raw_array_cohort_extract.py:2]: All the GATK code to support arrays was removed a while ago, I think we should nuke this file as well (and also from the docker build script)

File: scripts/variantstore/wdl/extract/utils.py (hunk 1 above)
Comment (kcibul) [reply scripts/variantstore/wdl/extract/utils.py:39]: Also… how id you decide on these two exceptions?  When this was originally (poorly) written it was also meant to handle rate limit and quota problems.  How are those surfaced and retried?

**Comment by kcibul** [comment]
Overall the refactoring looks good and makes sense… but I'm not seeing how this fixes the problem of eating exceptions we saw during a recent run.  Can you explain what was happening before, and how the new code addresses it?

**Comment by rsasch** [comment]
> Overall the refactoring looks good and makes sense… but I'm not seeing how this fixes the problem of eating exceptions we saw during a recent run. Can you explain what was happening before, and how the new code addresses it?

Sure!  This code (besides refactoring so that it was only in one place) aims to fix two issues:
1. if query results in an error, it gets run three more times and then, because of `while len(retry_delay) > 0`, it doesn’t run again and the `raise err` line never gets executed, so no error is ever raised
2.  if the query fails for a reason that has no chance of being fixed by a retry  (eg. 404), it will still run three more times

I probably missed some errors that should be "retry-able" (maybe `Aborted `? `BadGateway`? `Cancelled `? [full list here](https://googleapis.dev/python/google-api-core/latest/exceptions.html)), but I still think it makes sense to not treat all errors the same.

File: scripts/variantstore/wdl/extract/utils.py (hunk 2 above)
Comment (rsasch) [reply scripts/variantstore/wdl/extract/utils.py]: good catch — changed 👍🏻

File: scripts/variantstore/wdl/extract/raw_array_cohort_extract.py (hunk 3 above)
Comment (rsasch) [reply scripts/variantstore/wdl/extract/raw_array_cohort_extract.py:2]: I couldn't find any reference to it in the `Dockerfile` or `build_docker.sh`, but it was called by `raw_array_cohort_extract.wdl`?  Should the WDL file be deleted, too?

Comment (kcibul) [reply scripts/variantstore/wdl/extract/raw_array_cohort_extract.py:2]: yeah we should delete that arrays WDL also

**Review by kcibul — APPROVED** [review]
lgtm!

**Pull request merged**